core/client/src/blockchain.rs

```diff
@@ -195,56 +215,128 @@ impl<Block: BlockT> TreeRoute<Block> {
 	}
 }
 
+/// Get lowest common ancestor between two blocks in the tree.
+///
+/// Note: this implementation is efficient because of our current query pattern:
+/// lca(best, final) (O(h)), lca(best + 1, final) (O(1)), lca(best + 2, final) (O(1)), etc.
+pub fn lowest_common_ancestor<
```

Comment:
1. Why not just make it a client method and get rid of functor arguments? It does require the client in all cases anyway.

2. Do I get it right that in order to function correctly this requires the arguments to follow ancestry order between calls? Otherwise the `ancestor` field may end up skipping the actual ancestor required for another call. Is there a fallback that work even if this condition is not upheld?

Reply:
2. Yes, it's true, it has linear complexity in those cases (I'm missing something if you mean incorrect as with wrong result ^^). I thought about adding more links when you traverse a long path and about implementing a generic algorithm for LCA (maybe using something like https://www.topcoder.com/community/competitive-programming/tutorials/range-minimum-query-and-lowest-common-ancestor/), but I'm still not convinced is worth to lose the `O(1)` (in most cases) and simplicity. Do you have an idea?

Reply:
Also, the paths are not so long, and if they are long we would need to save something to database probably.

Reply:
Even in the case where our ancestor is too old and we fallback to linear search it should be faster than what we had previously because this linear search will very likely all be done in-memory (all hashes will most likely be cached because we add them when importing a block).